### PR TITLE
Gradually prune EVM state out of app.db

### PR DESCRIPTION
**Pull request by pathornteng**
Currently, EVM state has been moved out of app.db to evm.db but the keys have not been deleted yet. To further improve PlasmaChain performance, we should gradually remove VM keys from app.db. This PR gradually removes vm keys in every block commit. The number of deleted keys in each block must be carefully determined because too many delete operations could degrade chain performance. A feature flag (AppStore 3.1) might be used to activate this behavior.

Rref: https://github.com/loomnetwork/loomchain/issues/1178

- [ ] I added unit tests for any code that added
- [ ] I updated the CHANGELOG.md 
- [x] All IP is original and not copied from another source
- [x] I assign all copyright to Loom Network for the code in the pull request

### CODE REVIEW

File: store/multi_writer_app_store.go [hunk 1]
```diff
@@ -167,12 +206,30 @@ func (s *MultiWriterAppStore) SaveVersion() ([]byte, int64, error) {
...
 		// inconsistency in evm.db across the cluster
-		// AppStore 3.1 write EVM root to app.db only if it changes
+		// AppStore 3.1 writes EVM root to app.db only if it changes
 		if bytes.Equal(s.appStore.Get(appStoreVersion3_1), []byte{1}) {
 			oldRoot := s.appStore.Get(rootKey)
 			if !bytes.Equal(oldRoot, currentRoot) {
 				s.appStore.Set(rootKey, currentRoot)
 			}
+
+			// vm keys deletion process
```
Comment (enlight) [inline store/multi_writer_app_store.go]: Why is this nested in here? It's independent of the v3.1 / evm root stuff.

File: store/multi_writer_app_store.go [hunk 2]
```diff
@@ -93,10 +116,26 @@ func NewMultiWriterAppStore(appStore *IAVLStore, evmStore *EvmStore, saveEVMStat
...
+func (s *MultiWriterAppStore) loadOnChainConfig() {
+	configBytes := s.Get([]byte(configKey))
+	cfg := config.DefaultConfig()
+	s.numEvmKeysToPrune = int(cfg.AppStore.NumEvmKeysToPrune)
+	if len(configBytes) > 0 {
+		if err := proto.Unmarshal(configBytes, cfg); err != nil {
+			panic(err)
+		}
+	}
+	if int(cfg.AppStore.NumEvmKeysToPrune) > 0 {
```
Comment (enlight) [inline store/multi_writer_app_store.go]: Why are these `int` conversions needed?

File: store/multi_writer_app_store.go [hunk 3]
```diff
@@ -167,12 +206,30 @@ func (s *MultiWriterAppStore) SaveVersion() ([]byte, int64, error) {
...
+				begin := time.Now()
+				s.loadOnChainConfig()
+				rangeData := s.appStore.RangeWithLimit(vmPrefix, s.numEvmKeysToPrune)
+				for _, data := range rangeData {
+					s.appStore.Delete(util.PrefixKey(vmPrefix, data.Key))
+				}
+				// If rangeData is empty, it means all vm keys are deleted.
+				// So set the flag to stop the deletion
+				if len(rangeData) == 0 {
+					s.appStore.Set(evmStateDeletedKey, []byte{1})
```
[the remark this answers is not in the document]
Comment (pathornteng) [reply store/multi_writer_app_store.go]: Done

File: store/multi_writer_app_store.go [hunk 4]
```diff
@@ -93,10 +116,26 @@ func NewMultiWriterAppStore(appStore *IAVLStore, evmStore *EvmStore, saveEVMStat
 		store.onlySaveEvmStateToEvmStore = bytes.Equal(store.appStore.Get(evmDBFeatureKey), []byte{1})
 	}
 
+	store.evmStateDeleted = bytes.Equal(store.appStore.Get(evmStateDeletedKey), []byte{1})
```
[the remark this answers is not in the document]
Comment (pathornteng) [reply store/multi_writer_app_store.go]: Done

File: store/multi_writer_app_store.go [hunk 5]
```diff
@@ -167,12 +206,30 @@ func (s *MultiWriterAppStore) SaveVersion() ([]byte, int64, error) {
...
 			oldRoot := s.appStore.Get(rootKey)
 			if !bytes.Equal(oldRoot, currentRoot) {
 				s.appStore.Set(rootKey, currentRoot)
 			}
+
+			// vm keys deletion process
+			if bytes.Equal(s.appStore.Get(appStoreVersion3_2), []byte{1}) && !s.evmStateDeleted {
+				begin := time.Now()
+				s.loadOnChainConfig()
+				rangeData := s.appStore.RangeWithLimit(vmPrefix, s.numEvmKeysToPrune)
```
[the remark this answers is not in the document]
Comment (pathornteng) [reply store/multi_writer_app_store.go]: Done

File: store/multi_writer_app_store.go (hunk 1 above)
Comment (pathornteng) [reply store/multi_writer_app_store.go]: move it out

File: store/multi_writer_app_store.go (hunk 2 above)
Comment (pathornteng) [reply store/multi_writer_app_store.go]: because `RangeWithLimit` takes `int` but `NumEvmKeysToPrune` is `uint64`

**Pull request merged**